### PR TITLE
Add Automatic-Module-Name attribute to library, reflect, compiler

### PR DESCRIPTION
**Pull request by retronym**
Adds a `Automatic-Module-Name` entry to the `MANIFEST.MF` files of `scala-library` / `scala-reflect` and `scala-compiler`, see / fixes scala/bug#10734.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -400,6 +402,7 @@ lazy val compiler = configureAsSubproject(project)
   .settings(generatePropertiesFileSettings)
   .settings(generateBuildCharacterFileSettings)
   .settings(Osgi.settings)
+  .settings(AutomaticModuleName.settings("scala.tools.nsc"))
```
Comment (retronym) [inline build.sbt:405]: I wasn't too sure what to call this one, or if we should include the attribute at all to the current, monolithic scala-compiler.jar

File: project/AutomaticModuleName.scala [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ *
+ * The names carry a lot of implications and DO NOT have to always align 1:1 with the group ids or package names,
+ * though there should be of course a strong relationship between them.
+ */
+object AutomaticModuleName  {
+  def settings(name: String): Seq[Def.Setting[_]] = {
+    val pair = ("Automatic-Module-Name" -> name)
+    Seq(
+      packageOptions in (Compile, packageBin) += Package.ManifestAttributes(pair),
+      Osgi.headers += pair
```
Comment (retronym) [inline project/AutomaticModuleName.scala:19]: The second line is needed because of the way that the OSGI bundle task writes our JARs.

**Comment by retronym** [comment]
Manually tested:

```
⚡sbt setupPublishCore publishLocal
...
⚡ for x in library reflect compiler; do echo $x; jardiff /Users/jz/.ivy2/local/org.scala-lang/scala-$x/2.12.5-bin-dd16313-SNAPSHOT/jars/scala-$x.jar/META-INF/MANIFEST.MF | grep Auto; done
library
Automatic-Module-Name: scala.library
reflect
Automatic-Module-Name: scala.reflect
compiler
Automatic-Module-Name: scala.tools.nsc
```

File: build.sbt (hunk 1 above)
Comment (lrytz) [reply build.sbt:405]: @ktoso what do you think?

Comment (retronym) [reply build.sbt:405]: To elaborate, I thought "scala.compiler" was a bit too broad for something that is, in historical context, just one of the succession of Scala compilers, each of which ends up in different packages.

**Comment by lrytz** [comment]
Once merged, I could add a test to https://github.com/scala/scala-dist-smoketest/blob/2.12.x/scripts/jobs/release/smoketest#L73

File: build.sbt [hunk 3]
```diff
@@ -337,6 +337,7 @@ lazy val bootstrap = project in file("target/bootstrap")
 lazy val library = configureAsSubproject(project)
   .settings(generatePropertiesFileSettings)
   .settings(Osgi.settings)
+  .settings(AutomaticModuleName.settings("scala.library"))
```
Comment (ktoso) [inline build.sbt:340]: I guess that's the only one (and reflect I guess) that will show up in end-user module files right? "most" users don't care about the compiler module

File: build.sbt (hunk 1 above)
Comment (ktoso) [reply build.sbt:405]: If you think this name is good then ok. Note that there's no rule to stick to package names really, but this sounds probably well namespaced 👍

File: project/AutomaticModuleName.scala (hunk 2 above)
Comment (ktoso) [reply project/AutomaticModuleName.scala:19]: Looks good 👍

**Review by ktoso — APPROVED** [review]
Looks good

**Review by lrytz — APPROVED** [review]

**Pull request merged**